// ui/popup/bubble.js

const {DOMElement} = require("ui/dom/domelement.js")


class Bubble extends DOMElement {
    constructor(parent_element) {
        super("div")
        this.parentElement = parent_element
        this.setStyle({
            position: 'absolute',
            textAlign: 'center',
            backgroundColor: 'rgb(180,180,180)',
            border: '2px solid #505050',
            zIndex:10000,
            borderRadius: '8px'
        })
    }

    open(element) {
        let window_rect = this.parentElement.getBoundingClientRect()
        let element_rectangle = element.getBoundingClientRect()

        this.parentElement.appendChild(this.domElement)
        let my_rect = this.domElement.getBoundingClientRect()
        this.setStyle({
            top: (element_rectangle.bottom - window_rect.top) + "px",
            left: (element_rectangle.right - window_rect.left - (my_rect.width / 2)) + "px",
        })
    }

    close() {
        this.parentElement.removeChild(this.domElement)    
    }
}

module.exports = {
    Bubble: Bubble
}
