// ui/popup/tag_filter_picker.js
const { DOMElement } = require("ui/dom/domelement.js")
const { Bubble } = require("ui/popup/bubble.js")
// const { SortableListView } = require('ui/listview/sortable')
const { ClusteredListView } = require('ui/listview/cluster')
const { ScrollArea } = require("ui/dom/scroll_area.js")
const { Checkbox } = require("ui/dom/checkbox.js")
const { Label } = require("ui/dom/label.js")
const { ConfirmButton, DismissButton } = require("ui/dialog/buttons.js")
const { BoxLayout } = require("ui/layout/box")
const { TextRenderer, CheckboxRenderer } = require('ui/listview/renderer')


class TagListView extends ClusteredListView {
    constructor(tableBody, scrollerElement) {
        super(tableBody, scrollerElement)
        this.tableBody = tableBody
        this.tags = []
        this.checked = {}
        this.tagObjects = {}

        this.addColumnRenderer(new CheckboxRenderer({
            width:'20px',
            textAlign: "left",
            id: (rowId, columnId) => {
                return `${this._viewId}-filter-${rowId}`
            },
            checked: (obj) => {return obj.checked}
        }))

        this.addColumnRenderer(new TextRenderer({
            textAlign: "left",
            editTag: "name",
            color: "rgb(20,20,20)",
            text: (obj) => {return obj.name}
        }))
    }

    convert_tag(x) {
        let element = {
            id: x.id,
            checked: x.checked,
            name: x.name,
            color: x.color
        }
        return element
    }

    startReorder() {
        this._reordering = true
    }

    listReordered(new_order) {
        this._reordering = false

        this.tags = new_order.map((x) => {
            return this.tagObjects[this._elements.getElementById(x).id]
        })
    }

    handleClick(e) {
        let tableRow = e.target.closest("tr")
        let tagObject = this._elements.getElementById(tableRow.id)
        let eventElementId = e.target.id
        let color_regex = new RegExp(`${this._viewId}-filter-([_a-zA-Z0-9]+)`)
        let matches = color_regex.exec(eventElementId)
        if (matches != undefined) {
            if (e.target.checked) {
                this.checked[tagObject.id] = true
            } else {
                delete this.checked[tagObject.id]
            }
            return;
        }
    }

    setList(list_elements) {
        if (list_elements == undefined) {
            list_elements = []
        }
        this.tagObjects = {}
        list_elements.forEach((x) => {
            if (x.checked) {
                this.checked[x.id] = true                
            }
            this.tagObjects[x._id] = x
        })
        let rowIds = this.setRows(list_elements.map((e) => {
            return e 
        }))
        this.tags = list_elements.map((x) => { return x })
        this.tableBody.innerHTML = this._rendered_rows.getList().join("")
        let elements = document.querySelectorAll(`.${this._viewId}`);
        this.setRowElements(elements)
        this._objectid_to_row_id = {}

        Object.keys(rowIds).forEach((id) => {
            this._objectid_to_row_id[rowIds[id].id] = id
        })
    }

    // setChecked(list) {
    //     list.forEach((x) => {
    //         console.log(x)
    //     })
    // }

    getCheckedTags() {
        return this.checked
    }

}


class TagFilterPicker extends Bubble {
    constructor(parent_element, config) {
        super(parent_element)
        this.configuration = config

        this._filterUnplayed = this.configuration.filterUnplayed || false

        this.layout = new BoxLayout("vertical", {left: 10, right: 10, top: 10, bottom: 10, padding: 10})

        let unplayedFilterLayout = new BoxLayout("horizontal", {left: 0, right: 0, top: 0, bottom: 0, padding: 0})
        let unplayedFilterCheck = new Checkbox()
        unplayedFilterCheck.on("value-changed", (value) => {
            this._filterUnplayed = value
        })
        unplayedFilterCheck.setChecked(this._filterUnplayed)
        let unplayedFilterLabel = new Label("Only unplayed tracks")
        unplayedFilterLabel.setStyle({
            color: "rgb(20,20,20)"
        })
        unplayedFilterLayout.append(unplayedFilterCheck, false, true, 15)
        unplayedFilterLayout.append(unplayedFilterLabel, true, true)

        let tagListScroller = new ScrollArea()
        tagListScroller.setStyle({
            overflow: null,
            overflowY: "auto"
        })
        let tagListTable = new DOMElement("table")
        tagListTable.setStyle({
            width: "100%"
        })
        // let tagListBody = new DOMElement('table')
        // tagListTable.appendChild(tagListBody)
        this._input = new TagListView(tagListTable.domElement, tagListScroller.domElement)
        tagListScroller.appendChild(tagListTable)

        this._input.setList(this.configuration.tags || [])
        this.layout.append(unplayedFilterLayout, false, true, 20)
        this.layout.append(tagListScroller, true, true)

        let save_button = new ConfirmButton()
        save_button.setText("Filter")
        save_button.on("click", () => {
            if (this.configuration.updateFilter != undefined) {
                this.configuration.updateFilter({
                    filterUnplayed: this._filterUnplayed,
                    filterTags: this._input.getCheckedTags()
                })
                this.configuration.cancel && this.configuration.cancel()
                this.close()
            }
        })

        let dismiss_button = new DismissButton()
        dismiss_button.setText("Close")
        dismiss_button.on("click", () => {
            this.configuration.cancel && this.configuration.cancel()
            this.close()
        })

        let button_layout = new BoxLayout("horizontal", {padding: 20})
        button_layout.append(save_button, true, true)
        button_layout.append(dismiss_button, true, true)
        this.layout.append(button_layout, false, true, 30)


        this.appendChildren([this.layout])
        this.setStyle({width:'200px', height:'350px'})
    }

}

module.exports = {
    TagFilterPicker: TagFilterPicker
}
