// iface/dialogs/session_save_dialog.js
const { InputDialog } = require("ui/dialog/input")
const { DOMElement } = require("ui/dom/domelement.js")

// <dialog id="save-session-dialog" style="z-index:1000; padding:0; background-color:rgb(100,100,100)">
// <h1 style="background-color:rgb(30,30,30); text-align:center; font-size:11pt; margin:0px; color:rgb(200,200,200); height:30px; line-height:30px">Save current session</h1>
// <div style="margin:10px">
//     <input id="session-name" type="text" style="width:calc(100% - 40px); margin:2px; padding-left:10px; padding-right:10px; height:20px" placeholder="Name...">
//     <input id="session-location" type="text" style="width:calc(100% - 40px); margin:2px; padding-left:10px; padding-right:10px; height:20px" placeholder="Location...">
//     <input id="session-address" type="text" style="width:calc(100% - 40px); margin:2px; padding-left:10px; padding-right:10px; height:20px" placeholder="Address...">
// </div>
// <menu>
//     <button id="session-save">SAVE</button>
//     <button id="session-save-cancel">CANCEL</button>
// </menu>
// </dialog>



class SessionSaveDialog extends InputDialog {
    constructor(config) {

        super({
            confirmAction: config.confirmAction,
            dismissAction: config.dismissAction,
            confirmText: "save",
            dismissText: "cancel"
        })
        this.setTitle("Save the current session...")
        this.input_area.setStyle({
            padding:'10px'
        })
        this.name_input = this.makeInput("Name...")
        this.location_input = this.makeInput("Location...")
        this.address_input = this.makeInput("Address...")
        this.input_area.appendChildren([this.name_input, this.location_input, this.address_input])

    }

    makeInput(placeholder) {
        let new_input = new DOMElement("input")
        new_input.setStyle({
            width:'calc(100% - 40px)', 
            margin:'2px',
            paddingLeft:'10px', 
            paddingRight:'10px',
            height:'20px',
            color: "rgb(30,30,30)"
        })
        new_input.domElement.type = "text"
        new_input.domElement.placeholder = placeholder
        return new_input
    }

}

module.exports = {
    SessionSaveDialog: SessionSaveDialog
}
