// iface/dialogs/tag_edit_dialog.js
const { DOMElement } = require("ui/dom/domelement.js")
const { ColorPicker } = require("ui/popup/colorpicker.js")
const { ScrollArea } = require("ui/dom/scroll_area.js")
const { TextInput } = require("ui/dom/text_input.js")
const { Button } = require("ui/dom/button.js")
const { Dialog } = require("ui/dialog/dialog.js")
const { BoxLayout } = require("ui/layout/box")
const { ConfirmButton, DismissButton } = require("ui/dialog/buttons.js")
const { ClusteredListView } = require('ui/listview/cluster')
const { TextRenderer } = require('ui/listview/renderer')


class TagListView extends ClusteredListView {
    constructor(model, body, scroller) {
        super(body, scroller)

        this.setupColumnRenderers()
        this._objectid_to_row_id = {}
        this.model = model
    }

    convert_tag(tag_element) {
        return {
            id: tag_element._id,
            name: tag_element.name,
            color: tag_element.color
        }
    }

    setDimmedRows(idList) {
        let rowIdList = idList.map((x) => {
            return this._objectid_to_row_id[x]
        }).filter((x) => {
            return x
        })
        super.setDimmedRows(rowIdList)
    }


    setupColumnRenderers() {
        this.addColumnRenderer({
            _config: {}, 
            render: (row, col, tag) => {
                return `<td  id='${this._viewId}-color-${tag.id}' style="width:25px; text-align:center">
                    <button class="main-list color-chooser show-color-picker" style="background-color:${tag.color}"></button>
                </td>`
        }})

        this.addColumnRenderer(new TextRenderer({
            width: "250px",
            editTag: "name",
            color: (obj) => {return "rgb(28,28,28)"},
            text: (obj) => {return obj.name}
        }))

        this.addColumnRenderer(new TextRenderer({
            width: "50px",
            textAlign: "right",
            id: (rowId, columnId) => {
                return `${this._viewId}-edit-${rowId}`
            },
            color: (obj) => {return "rgb(100,100,100)"},
            text: (obj) => {return `<i class="fa fa-edit"></i>`}
        }))

        this.addColumnRenderer(new TextRenderer({
            width: "25px",
            textAlign:"center",
            id: (rowId, columnId) => {
                return `${this._viewId}-remove-${rowId}`
            },
            color: (obj) => {return "rgb(158,28,28)"},
            text: (obj) => {return `<i class="fa fa-trash"></i>`}
        }))
    }


    async handleEditKeypress(e) {
        let new_name = this.editValues['name'].value
        if ((new_name == this.oldName) || (this.model.checkNameAvailability(new_name))) {
            this.editValues['name'].style.color = null
        } else {
            this.editValues['name'].style.color = "#dd0000"
        }
        e.preventDefault()
    }

    beginEdit(id) {
        let rowId = super.beginEdit(id)
        this.editId = this._elements.getElementById(rowId).id
        this.oldName = this._elements.getElementById(rowId).name
    }

    async saveEdit() {
        let values = super.saveEdit()
        this.dispatch("tag-name-edited", {
            _id: this.editId,
            name: values.name
        })
    }

    handleClick(e) {
        let tableRow = e.target.closest("tr")
        let trackObject = this._elements.getElementById(tableRow.id)
        let eventElementId = e.target.closest("td").id
        let rating_regex = new RegExp(`${this._viewId}-edit-([_a-zA-Z0-9]+)`)
        let matches = rating_regex.exec(eventElementId)
        if (matches != undefined) {
            this.beginEdit(tableRow.id)
            e.preventDefault()
            return;
        }

        let loved_regex = new RegExp(`${this._viewId}-remove-([_a-zA-Z0-9]+)`)
        matches = loved_regex.exec(eventElementId)
        if (matches != undefined) {
            this.dispatch("tag-remove-clicked", trackObject.id)
            e.preventDefault()
            return;
        }

        let color_regex = new RegExp(`${this._viewId}-color-([_a-zA-Z0-9]+)`)
        matches = color_regex.exec(eventElementId)
        if (matches != undefined) {
            this.dispatch('tag-color-clicked', trackObject.id, e)
            e.preventDefault()
            return;
        }
    }

    compare_tags(a, b) {
        if (!a.name) { return -1 }
        if (!b.name) { return 1 }
        let x = a.name.toLowerCase();
        let y = b.name.toLowerCase();
        if (x < y) { return -1; }
        if (x > y) { return 1; }
        return 0;        
    }    

    setList(list_elements) {
        this.model = list_elements
        if (list_elements == undefined) {
            list_elements = []
        }

        list_elements.sort(this.compare_tags) 

        let rowIds = this.setRows(list_elements.map((e) => {
            return this.convert_tag(e)
        }))

        this._objectid_to_row_id = {}

        Object.keys(rowIds).forEach((id) => {
            this._objectid_to_row_id[rowIds[id].id] = id
        })
    }

    updateElement(tagObject) {
        let element = this.convert_tag(tagObject)
        let rowId = this._objectid_to_row_id[element.id]
        super.updateElement(rowId, element)
        let row = this._row_elements.getElementById(rowId)
        if (row) {
            row.innerHTML = this.renderRow(rowId, element)
        }
    }
}




class TagEditDialog extends Dialog {
    constructor(library, config) {
        super()
        this.setTitle("Edit tags")
        this.setStyle({width:'400px', height:"500px"})
        this.library = library
        this._tagsList = []
        this._tagNames = {}
        this._tagObjects = {}
        this._tagsToRemove = {}
        this._tagsToAdd = {}
        this._tagsToUpdate = {}

        this.layout = new BoxLayout("vertical", {left:20, right:20, top:20, bottom:20, padding:10})
        this.layout.setStyle({
            position: "absolute",
            top: '0px',
            width: "100%",
            height: "100%"
        })

        let text = new DOMElement("div")
        text.setStyle({
            textAlign: "justify",
            margin: "20px",
            color:"rgb(30,30,30)"
        })
        text.domElement.innerHTML = "Edit, add or remove tags.  Deleting a tag will remove it from all tracks it was assigned to."

        let newTagColor = null
        let newTagColorInput = new Button()
        newTagColorInput.on("click", () => {
            let colorPicker = new ColorPicker(document.getElementsByTagName("body")[0], {
                cancel: () => {
                    colorPicker.close()
                },
                chooseColor: (color) => {
                    newTagColor = color
                    newTagColorInput.domElement.style.backgroundColor = newTagColor
                    colorPicker.close()
                }                    
            })
            colorPicker.open(newTagColorInput.domElement)
        })
        let newTagNameInput = new TextInput()
        let addNewTagButton = new Button()
        addNewTagButton.setText("Create...")
        addNewTagButton.on("click", () => {
            if (newTagNameInput.domElement.value != "") {
                if (this.checkNameAvailability(newTagNameInput.domElement.value)) {
                    let newId = this._id()
                    let newTag = {
                        _id: newId,
                        color: newTagColor,
                        name: newTagNameInput.domElement.value,
                    }
                    this._tagObjects[newId] = newTag
                    this._tagsToAdd[newId] = newTag
                    this._tagsList.push(newTag)
                    this._listview.setList(this._tagsList)
                    newTagColor = null
                    newTagNameInput.domElement.value = ""
                    newTagColorInput.domElement.style.backgroundColor = newTagColor                        
                }
            }
        })

        let newTagInput = new BoxLayout('horizontal')
        newTagInput.append(newTagColorInput, false, true, 25)
        newTagInput.append(newTagNameInput, true, true)
        newTagInput.append(addNewTagButton, false, true, 70)

        let save_button = new ConfirmButton()
        save_button.setText("Save")
        save_button.on("click", () => {
            Object.values(this._tagsToAdd).filter((x) => {
                return !(this._tagsToRemove[x._id])
            }).forEach(async (newTag) => {
                await this.library.createTag(newTag.name, newTag.color)
            })
            Object.keys(this._tagsToRemove).filter((k) => {
                return this._tagsToRemove[k]
            }).forEach(async (k) => {
                await this.library.removeTag(k)
            })
            Object.values(this._tagsToUpdate).forEach(async (k) => {
                await this.library.updateTag(k)
            })
            this.close()
        })

        let dismiss_button = new DismissButton()
        dismiss_button.setText("Close")
        dismiss_button.on("click", () => {
            this.close()
        })

        let button_layout = new BoxLayout("horizontal", {padding: 20})
        button_layout.append(save_button, true, true)
        button_layout.append(dismiss_button, true, true)
        

        let tagListScroller = new ScrollArea()
        tagListScroller.setStyle({
            overflow: null,
            overflowY: "auto"
        })
        let tagListTable = new DOMElement("table")
        let tagListBody = new DOMElement('tbody')
        tagListTable.appendChild(tagListBody)
        this._listview = new TagListView(this, tagListBody.domElement, tagListScroller.domElement)
        tagListScroller.appendChild(tagListTable)


        this._listview.on("tag-color-clicked", (tagId, e) => {
            let colorPicker = new ColorPicker(this.domElement, {
                cancel: () => {
                    colorPicker.close()
                },
                chooseColor: (color) => {
                    this.updateValues({
                        _id: tagId,
                        color: color
                    })
                    colorPicker.close()
                }                    
            })
            colorPicker.open(e.target)
        })

        this._listview.on("tag-name-edited", (newValues) => {
            this.updateValues(newValues)
        })

        this._listview.on("tag-remove-clicked", (id) => {
            if (this._tagsToRemove[id]) {
                this._tagsToRemove[id] = false
            } else {
                this._tagsToRemove[id] = true
            }
            this._listview.setDimmedRows(Object.keys(this._tagsToRemove).filter((x) => {
                return this._tagsToRemove[x]
            }))

        })

        this.layout.append(text, false, false)
        this.layout.append(newTagInput, false, true, 30)
        this.layout.append(tagListScroller, true, true)
        this.layout.append(button_layout, false, true, 50)
        this.appendChildren([this.layout])
    }

    checkNameAvailability(newName) {
        console.log((this._tagNames[newName]))
        return (!(this._tagNames[newName]) && true) || false
    }

    updateValues(newValues) {
        if (this._tagsToUpdate[newValues._id]) {
            this._tagsToUpdate[newValues._id].name = newValues.name || this._tagsToUpdate[newValues._id].name
            this._tagsToUpdate[newValues._id].color = newValues.color || this._tagsToUpdate[newValues._id].color
        } else if (this._tagObjects[newValues._id] && !(this._tagsToAdd[newValues._id])){
            this._tagsToUpdate[newValues._id] = {
                _id: newValues._id,
                name: newValues.name || this._tagObjects[newValues._id].name,
                color: newValues.color || this._tagObjects[newValues._id].color
            }
        } else if (this._tagsToAdd[newValues._id]) {
            this._tagsToAdd[newValues._id].name = newValues.name || this._tagsToAdd[newValues._id].name
            this._tagsToAdd[newValues._id].color = newValues.color || this._tagsToAdd[newValues._id].color
        } 

        if (this._tagObjects[newValues._id]) {
            this._tagObjects[newValues._id].name = newValues.name || this._tagObjects[newValues._id].name
            this._tagObjects[newValues._id].color = newValues.color || this._tagObjects[newValues._id].color
        }
        this._listview.updateElement(this._tagObjects[newValues._id])
    }

    async init() {
        this._tagsList = Object.values(await MDB.tags.getAllObjects())
        this._listview.setList(this._tagsList)
    }

    async open() {
        super.open()
    }
}

module.exports = {
    TagEditDialog: TagEditDialog
}
