// track_playback/remote_player/window/playback.js
const {AudioTrackPlayer} = require('track_playback/track_player.js')


class PlaybackController extends AudioTrackPlayer {
    constructor() {
        super()
        this.playing = false

        this.monitor_set_volume = 1;
        this.monitor_volume = 1;
        this.monitor_muted_volume = 0;
        this._lastPos = undefined


        this.on('end-of-stream', () => {
            ipcRenderer.sendTo(parentId, `${response_prefix}end-of-stream`, {})
        })

        this.on('playback-stopped',  () => {
            ipcRenderer.sendTo(parentId, `${response_prefix}playback-stopped`, {})
        })
    
        this.on('playback-paused', () => {
            ipcRenderer.sendTo(parentId, `${response_prefix}playback-paused`, {})
        })
        
        this.on('playback-started', () => {
            ipcRenderer.sendTo(parentId, `${response_prefix}playback-started`, {})
        })

        this.on('stream-position', (pos) => {
            if (this._lastPos && (pos - this._lastPos < 100)) {
                return
            } else {
                this._lastPos = pos
            }
            ipcRenderer.sendTo(parentId, `${response_prefix}stream-position`, {
                position: pos, 
                duration: this.source.duration
            })
        })
    }
}

module.exports = {
    PlaybackController: PlaybackController
}
