// musicdb/collection.js
const { EventDispatcher } = require("event_dispatcher")
const { Datastore } = require("nedb-async-await")


class CollectionController extends EventDispatcher {
    constructor(filename) {
        this.d = new Datastore({
            filename: filename, 
            autoload: true,
            timestampData: true})
    }

    getAll() {
        return this.getIds()
    }

    getIds(id_array) {
        if (id_array != undefined) {
            return this.d.find({"_id": {"$in": id_array}})
        } else {
            return this.d.find({})
        }
    }

    async getId(id) {
        return (await this.getIds([id]))[0]
    }

    async length() {
        return await this.d.count({})        
    }

    async setData(object, new_data) {
        let [num, modified] = await this.d.update({_id: object._id},  {$set: new_data}, {returnUpdatedDocs: true})
        if (num > 0) {
            this.dispatch("track-metadata-changed". modified)
        }
        return modified
    }

}


class TrackCollectionController extends CollectionController {
    constructor(filename) {
        super(filename)
    }

    async duration(id_array) {
        let tracks = await this.getIds(id_array)
        let dur = 0
        tracks.forEach((track) => {
            dur += (track.bounds.end - track.bounds.start)
        })
        return dur
    }
}


module.exports = {
    CollectionController: CollectionController,
    TrackCollectionController: TrackCollectionController
}
