// musicdb/playlist.js
const { TrackSetController } = require("./object_set.js")


class PlaylistController extends TrackSetController {
    constructor(db, collection, id) {
        super(db, collection, id)
    }

    compareTracks(a, b) {
        let x = a.title.toLowerCase();
        let y = b.title.toLowerCase();
        if (x < y) {return -1;}
        if (x > y) {return 1;}
        return 0;        
    }


    // async getAll() {
    //     let e = await this.c.d.find(this.q)
    //     e.sort(this.compareTracks)
    //     return e
    // }

}


class PlaylistViewController extends TrackSetController {
    constructor(db, collection, id) {
        super(db, collection, id)

    }

    compareTracks(a, b) {
        let x = a.title.toLowerCase();
        let y = b.title.toLowerCase();
        if (x < y) {return -1;}
        if (x > y) {return 1;}
        return 0;        
    }

    getId(id) {
        return this.d.getTrackById(id)
    }
    // async getAll() {
    //     let e = await this.c.d.find(this.q)
    //     e.sort(this.compareTracks)
    //     return e
    // }

}


class QueryController extends TrackSetController {
    constructor(db, collection, q) {
        super(db, collection)
        this.q = q
    }

    compareTracks(a, b) {
        let x = a.title.toLowerCase();
        let y = b.title.toLowerCase();
        if (x < y) {return -1;}
        if (x > y) {return 1;}
        return 0;        
    }


    async getAll() {
        let e = await this.c.d.find(this.q)
        e.sort(this.compareTracks)
        return e
    }

}

module.exports = {
    PlaylistController: PlaylistController,
    QueryController: QueryController,
    PlaylistViewController: PlaylistViewController
}
