// ui/listview/cluster.js
const Clusterize = require("clusterize.js")
const { BaseListView } = require("./base")

class ClusteredListView extends BaseListView {
    constructor(element, list_scroller) {
        super(element, list_scroller)
        this._cluster = new Clusterize({
            rows: [],
            scrollElem: this.scrollerElement,
            contentId: this.domElement,
            callbacks: {
                clusterWillChange: () => {
                    this.table_rows = {}
                },
                clusterChanged: () => {
                    let elements = document.querySelectorAll('.track-entry');
                    [].forEach.call(elements, (e) => {
                        let track_id = e.attributes["data-track-id"].value
                        this.table_rows[track_id] = e
                    });
                    this.setDimmedRows(Object.keys(this._dimmed))
                },
                scrollingProgress: (progress) => {}
            }
        });
    }
}

module.exports = {
    ClusteredListView: ClusteredListView
}
