// ui/listview/cluster.js
const Clusterize = require("clusterize.js")
const { BaseListView } = require("./base")

class ClusteredListView extends BaseListView {
    constructor(element, list_scroller) {
        super(element, list_scroller)
        this._cluster = new Clusterize({
            rows: [],
            scrollElem: this.scrollerElement,
            contentElem: this.domElement,
            callbacks: {
                clusterWillChange: () => {
                    this._row_elements.clear()
                },
                clusterChanged: () => {
                    let elements = document.querySelectorAll(`.${this._viewId}`);
                    this.setRowElements(elements)
                    this.setDimmedRows(Object.keys(this._dimmed))
                    this.clusterChanged(elements)
                },
                scrollingProgress: (progress) => {

                }
            }
        });
    }

    setRows(elems) {
        super.setRows(elems)
        // console.log(this._rows.getList())
        this._cluster.update(this._rows.getList())
    }
}

module.exports = {
    ClusteredListView: ClusteredListView
}
