// ui/listview/sortable.js
const Sortable = require("../../../lib/Sortable.js")
const { BaseListView } = require("./base")

class SortableListView extends BaseListView {
    constructor(element, list_scroller) {
        super(element, list_scroller)
        this.sortable = Sortable.create(this.list_dom,
            {
                animation: 150,
                ghostClass: "ghost",
                dataIdAttr: "data-track-id",
                onStart: (evt) => {

                },
                onEnd: async (evt) => {
                    let new_order = []
                    for(let i=0; i<this.list_dom.rows.length; i++) {
                        if (this.list_dom.rows[i].attributes["data-track-id"].value == 'null') {
                            new_order.push(null)
                        } else {
                            new_order.push((this.list_dom.rows[i].attributes["data-track-id"].value))
                        }
                    }
                    this.view_list_order = new_order
                    await this.controller.reorder_queue(this.view_list_order)

                    this.num_tracks_dom.innerHTML = `${await this.controller.q_length()} tracks`
                    this.duration_dom.innerHTML = `${format_seconds_long(Math.round(await this.controller.duration() / 1000))}`
                },
            }
        )
    }
}


module.exports = {
    SortableListView: SortableListView
}
