// ui/listview/sortable.js
const Sortable = require("../../../../lib/Sortable.js")
const { BaseListView } = require("./base")




class SortableListView extends BaseListView {
    constructor(scrollerElement, listBody) {
        super(listBody, scrollerElement)
        this.listBody = listBody
        this.sortable = Sortable.create(listBody,
            {
                animation: 150,
                dataIdAttr: "id",
                ghostClass: "ghost",

                onStart: (evt) => {
                    this.startReorder()
                },
                onEnd: (evt) => {
                    let new_order = []
                    for(let i=0; i<this.listBody.rows.length; i++) {
                        if (this.listBody.rows[i].id == 'null') {
                            new_order.push(null)
                        } else {
                            new_order.push((this.listBody.rows[i].id))
                        }
                    }
                    this.listReordered(new_order)
                    this._rendered_rows.reorder(new_order)
                }
            }
        )
    }

    listReordered(newOrder) {

    }

    startReorder() {

    }

    disable() {
        this.sortable.option("disabled", true);
    }
    
    enable() {
        this.sortable.option("disabled", false);
    }
}


module.exports = {
    SortableListView: SortableListView
}
