// app/views/header_main_player.js
const WaveSurfer = require("wavesurfer.js")
const WaveSurferRegions = require('wavesurfer.js/dist/plugin/wavesurfer.regions.min.js');
const moment = require('moment');

const { Image } = require("ui/dom/image.js")
const { Label } = require("ui/dom/label")
const { Rating } = require("app/components/rating")
const { CustomCheckbox } = require("app/components/favorite")
const { BoxLayout } = require("ui/layout/box")
const { DOMElement } = require("ui/dom/domelement.js")


class LabeledInfo extends BoxLayout {
    constructor(label) {
        super("vertical", {left:0, right:0, top:0, bottom:0, padding:0})
        this.label = new Label(label)
        this.label.setStyle({
            fontSize: "8pt",
            textAlign: "center",
            textTransform: "uppercase",
            color: "rgb(200,200,200)"
        })
        this.input = new Label()
        this.input.domElement.classList.remove('main-track-search')
        this.input.setStyle({
            marginTop: '8px',
            paddingLeft:'15px',
            textAlign: "center",
            color: 'rgb(200,200,200)',
            // backgroundColor: "rgb(190,190,190)",
            fontSize: '10pt',
            border: 'none',
            outline: "none"
        })
        this.append(this.label, true, true)
        this.append(this.input, true, true)
    }

    setText(v) {
        this.input.setText(v)
    }

    getText() {
        return this.input.getText()
    }
}



class MainPlayerView extends BoxLayout {
    constructor (track_list_model) {
        super('horizontal', {left: 0, right: 0, top: 0, bottom: 0, padding: 0})
        this.track_list_model = track_list_model
        this.controller = undefined

        this.c = document.getElementById("main-player")

        this.precueMenuOptions = new DropdownMenu()
        this.trackTitleLabel = new Label() 

        this.trackTitleLabel.setStyle({
            fontWeight: 'bold'
        })
        this.trackAlbumLabel = new Label() 
        this.trackAlbumLabel.setStyle({
            overflow: "hidden",
            textOverflow: "ellipsis",
            whiteSpace: 'nowrap'

        })
        this.trackArtistLabel = new Label()
        this.trackArtistLabel.setStyle({
            fontStyle: 'italic',
            whiteSpace: 'nowrap'
        })
        this.timeRemainingLabel = new Label()
        this.trackYearLabel = new LabeledInfo('year')
        this.trackGenreLabel = new LabeledInfo('genre')
        this.trackBPMLabel = new LabeledInfo("bpm")
        this.trackDurationLabel = new LabeledInfo('length')
        this.trackCoverImage = new Image()
        this.trackRating = new Rating("rgb(200,200,200)")
        this.trackLoved = new CustomCheckbox("rgb(200,200,200)")

        let playerLayout = new BoxLayout('horizontal', {padding: 5})
        playerLayout.append(this.trackCoverImage, false, true, 100)

        let infoLayout =  new BoxLayout('vertical', {padding: 2})
        let titleLayout =  new BoxLayout('vertical', {top:7, bottom:2, padding: 3})
        titleLayout.append(this.trackTitleLabel, false, true, 20)

        let artistLayout =  new BoxLayout('horizontal', {padding: 7})
        artistLayout.append(this.trackArtistLabel, true, true, 15)
        artistLayout.append(new Label("|"), false, true, 15)
        artistLayout.append(this.trackAlbumLabel, true, true, 15)
        titleLayout.append(artistLayout, false, true, 20)

        let metadataLayout = new BoxLayout('horizontal', {padding: 7})
        metadataLayout.append(this.trackDurationLabel, true, true)
        metadataLayout.append(this.trackBPMLabel, true, true)
        metadataLayout.append(this.trackYearLabel, true, true)
        metadataLayout.append(this.trackGenreLabel, true, true)
        metadataLayout.append(this.trackRating, false, true, 75)
        metadataLayout.append(this.trackLoved, false, true, 15)
        infoLayout.append(metadataLayout, false, true, 35)
        titleLayout.append(infoLayout, true, true)

        playerLayout.append(titleLayout, true, true, 55)
        playerLayout.append(this.timeRemainingLabel, false, true, 55)
        this.append(new DOMElement('div'), true, true)
        this.append(playerLayout, true, true, 75)
        // this.append(titleLayout, true, true, 55)
        // this.append(this.timeRemainingLabel, false, true, 55)
        this.append(new DOMElement('div'), true, true)

        this.c.appendChild(this.domElement)

        this.trackRating.on('changed', (value) => {
            this.updateRating(value)
        })

        this.trackLoved.on("changed", (value) => {
            this.updateLoved(value)
        })

        this.track_list_model.on("object-updated", (track) => {
            if (this._track != undefined) {
                if (track._id == this._track._id) {
                    this.update_track(track)
                }    
            }
        })
    }

    set_controller(controller) {
        this.controller = controller
        this.controller.on("main:stream-position", (pos) => {
            let remaining = Math.abs(pos.duration*1000 - pos.position)
            document.getElementById("main-player-time-remaining").innerHTML = `-${moment(remaining).format("m:ss")}`
        })
        // this.controller.on("main:queue-stopped", this.set_queue.bind(this))
        // // this.controller.on("main:track-finished", this.set_queue.bind(this))
        this.controller.on("main:track-start-request", this.set_track.bind(this))
        // this.controller.on("main:queue-finished", this.set_queue.bind(this))
        // this.controller.on("main:next-track-countdown", this.set_queue.bind(this))
    }

    update_track(track) {
        let stream_length = (track.track.stream_end - track.track.stream_start);
        this._track = track
        this.trackTitleLabel.setText(track.metadata.title)
        this.trackAlbumLabel.setText(track.metadata.album)
        this.trackArtistLabel.setText(track.metadata.artist)
        this.trackGenreLabel.setText(track.metadata.genre)
        this.trackYearLabel.setText(track.metadata.year)
        this.trackBPMLabel.setText(track.track.bpm)
        this.trackDurationLabel.setText(`${moment(stream_length).format("m:ss")}`)
        this.trackRating.setRating(track.stats.rating)
        this.trackLoved.setValue(track.stats.loved)
        let cover_source = undefined
        if (track.metadata.cover == null) {
            cover_source = "../../resources/images/default_album_cover.png"
        } else {
            cover_source = `file://${track.metadata.cover.small}`;
        }
        this.trackCoverImage.setImage(cover_source)
    }

    set_track(tr) {
        this.update_track(tr.track_object)
    }

    updateRating(new_value) {
        if (this.track_list_model != undefined) {
            this.track_list_model.setTrackMetadata(this._track, {'stats.rating': new_value})
        }
    }

    updateLoved(new_value) {
        if (this.track_list_model != undefined) {
            this.track_list_model.setTrackMetadata(this._track, {'stats.loved': new_value})
        }
    }

    init () {

    }
}


module.exports = {
    MainPlayerView: MainPlayerView
}
