// app/views/main_player_view.js
const WaveSurfer = require("wavesurfer.js")
const WaveSurferRegions = require('wavesurfer.js/dist/plugin/wavesurfer.regions.min.js');


class MainPlayerView extends EventDispatcher {
    constructor (track_list_model) {
        super()
        this.controller = undefined
        this.track_list_model = track_list_model

        if (this.track_list_model != undefined) {
            this.track_list_model.on("object-updated", (track) => {
                if (this._track != undefined) {
                    if (track._id == this._track._id) {
                        this.set_track_metadata(track)
                    }
                }
            })
        }

        document.getElementById("main-player-loved").addEventListener("click", () => {
            if (this._track != undefined) {
                this.updateLoved(!(this._track.stats.loved))
            }
        })
    }

    set_controller(controller) {
        this.controller = controller
        this.controller.on("main:stream-position", (pos) => {
            let remaining = Math.abs(pos.duration*1000 - pos.position)
            document.getElementById("main-player-time-remaining").innerHTML = `-${format_nanoseconds(remaining)}`
        })
        this.controller.on("main:queue-stopped", this.set_queue.bind(this))
        this.controller.on("main:track-finished", this.set_queue.bind(this))
        this.controller.on("main:track-start-request", this.set_track.bind(this))
        this.controller.on("main:queue-finished", this.set_queue.bind(this))
        this.controller.on("main:next-track-countdown", this.set_queue.bind(this))
    }

    set_queue() {

    }

    set_track_metadata(track) {
        if (this._track != undefined && track._id == this._track._id) {
            let stream_length = (track.track.stream_end - track.track.stream_start);
            document.getElementById("main-player-track-title").innerHTML = track.metadata.title
            document.getElementById("main-player-track-album").innerHTML = track.metadata.album
            document.getElementById("main-player-track-artist").innerHTML = track.metadata.artist
            document.getElementById("main-player-track-bpm").innerHTML = track.track.bpm
            document.getElementById("main-player-track-duration").innerHTML = `${format_nanoseconds(stream_length)}`
            this.setRating(track.stats.rating)
            this.setLoved(track.stats.loved)
            let cover_source = undefined
            if (track.cover == null) {
                cover_source = "../../resources/images/default_album_cover.png"
            } else {
                cover_source = `file://${track.metadata.cover.medium}`;
            }
            document.getElementById("main-player-track-cover").src = cover_source    
        }
    }

    set_track(track) {
        let file_name = track.track.path
        this._track = track
        this.set_track_metadata(track)
        this._waveform.load(file_name)
    }

    init() {
        this._waveform = WaveSurfer.create({
            container:     `#main-player-waveform`,
            waveColor:     'violet',
            progressColor: 'purple',
            height:        40,
            barHeight:     1,
            plugins: [
                WaveSurferRegions.create({
                    container: `#${this.main_waveform_id}`,
                    deferInit: false,
                })
            ]
        });
        this._position_tracker = this.controller.on("main:stream-position",
            (pos) => {
                let p = pos.position / this._track.track.duration
                p = Math.max(p,0.0)
                p = Math.min(p,1.0)
                this._waveform.seekAndCenter(p)
            }
        )
        this._waveform.on(
            "ready", () => {
                this._waveform.zoom(0)
                this.controller.startMainPlayback(this._track)
            }
        )
    }

    setRating (num) {
        var html = "";
        for (var i=1; i<6; i++) {
            html+=`<i id='main-rating-star-${i}' class='fa ` + ( i <= num ? "fa-star" : "fa-star-o") +`' style='margin-left:3px'></i>`;
        }
        document.getElementById("main-player-rating").innerHTML = html
        for (let i=1; i<6; i++) {
            document.getElementById(`main-rating-star-${i}`).addEventListener('click', () => {
                if (i == 1 && this._track.stats.rating == 1) {
                    this.updateRating(0)
                } else {
                    this.updateRating(i)
                }
            })
        }
    }

    setLoved (value){
        var html = "";
        this._track.stats.loved = value
        html+="<i title='"+value+"' class='fa " + (value ? "fa-heart" : "fa-heart-o") +"'></i>";
        document.getElementById("main-player-loved").innerHTML = html
    }

    updateRating(new_value) {
        if (this.track_list_model != undefined) {
            this.track_list_model.setTrackMetadata(this._track, {"stats.rating":new_value})
        }
    }

    updateLoved(new_value) {
        if (this.track_list_model != undefined) {
            this.track_list_model.setTrackMetadata(this._track, {"stats.loved": new_value})
        }
    }

}

module.exports = {
    MainPlayerView: MainPlayerView
}
