// app/views/sidebar_sessions_list.js
const { ClusteredListView } = require('ui/listview/cluster')
const { ToggleRenderer, RatingRenderer, TextRenderer, NumberRenderer, DateRenderer} = require('ui/listview/renderer')
const $ = require("jquery")


class SessionsListView extends ClusteredListView {
    constructor() {
        super(document.getElementById("session-list-elements"), 
              document.getElementById("sessions-list-scroller"))
        this.prevWidth = [];
        this.setupColumnRenderers()
        this._objectid_to_row_id = {}

        this.forward_content_changed = async (q) => {
            await this.updateContents(q)
        }
    }

    convert_session(track_element) {
        return {
            id:   track_element._id,
            name: track_element.event,
            data: track_element,
            date: track_element.date_start,
        }
    }

    setupColumnRenderers() {
        this.addColumnRenderer(new TextRenderer({
            width: "30px",
            color: (obj) => {return null},
            text: (obj) => {return ""}
        }))

        this.addColumnRenderer(new TextRenderer({
            width: "25px",
            color: (obj) => {return null},
            text: (obj) => {return `<i class="fa fa-circle-o arrow-icon"></i>`}
        }))
            
        this.addColumnRenderer(new TextRenderer({
            width: "250px",
            color: (obj) => {return null},
            text: (obj) => {return obj.name}
        }))

        this.addColumnRenderer(new DateRenderer({
            width: "100px",
            dateFormat: "MM-DD-YYYY",
            color: (obj) => {return "rgb(128,128,128)"},
            date: (obj) => {return obj.date}
        }))

        this.addColumnRenderer(new TextRenderer({
            width: "25px",
            color: (obj) => {return null},
            text: (obj) => {return `<i class="fa fa-angle-right"></i>`}
        }))
    }

    async updateContents(q) {
        this.setList(q)
    }

    async displayModel(model) {
        if (this.model != undefined) {
            this.model.un("content-changed", this.forward_content_changed) 
        }
        if (model) {
            this.model = model
            this.model.on("content-changed", this.forward_content_changed) 
            this.updateContents(Object.values(await this.model.getAllObjects()))    
        } else {
            this.setList([])
        }
    }


    setList(list_elements) {
        console.log(list_elements)
        console.log(this.model)
        if (list_elements == undefined) {
            list_elements = []
        }

        list_elements.sort((a, b) => {return (b.date_start - a.date_start)})

        let rowIds = this.setRows(list_elements.map((e) => {
            return this.convert_session(e)
        }))

        this._objectid_to_row_id = {}

        Object.keys(rowIds).forEach((id) => {
            this._objectid_to_row_id[rowIds[id].id] = id
        })
    }

    handleContextMenu(e) {
        let tableRow = e.target.closest("tr")
        let trackObject = this._elements.getElementById(tableRow.id)
        this.dispatch("context-menu", trackObject.id)
    }

    handleDragStart(e) {
        let tableRow = e.target.closest("tr")
        let trackObject = this._elements.getElementById(tableRow.id)
        e.dataTransfer.setData("text/plain", trackObject.id)
    }

    handleDoubleClick(e) {
        let tableRow = e.target.closest("tr")
        let trackObject = this._elements.getElementById(tableRow.id)
        this.dispatch("row-double-click", trackObject.id)
    }

    handleClick(e) {
        let tableRow = e.target.closest("tr")
        let trackObject = this._elements.getElementById(tableRow.id)
        this.dispatch("row-click", trackObject.id)
        this.selectRow(tableRow.id)
    }

    handleEditKeypress(e) {

    }

    beginEdit() {
        let rowId = super.beginEdit()
        this.editTrackId = this._elements.getElementById(rowId).id
    }

    saveEdit() {
        this.dispatch("row-values-edited", this.editTrackId, super.saveEdit())
    }

    updateElement(trackObject) {
        let element = this.convert_track(trackObject)
        let rowId = this._objectid_to_row_id[element.id]
        super.updateElement(rowId, element)
        let row = this._row_elements.getElementById(rowId)
        if (row) {
            row.innerHTML = this.renderRow(rowId, element)
        }
    }

    getSelection() {
        return super.getSelection().map((x) => {
            return this._elements.getElementById(x).id
        })
    }
}


module.exports = {
    SessionsListView: SessionsListView
}
