// app/views/sidebar.js
const { LinksList } = require('app/views/sidebar_library_links')
const { SessionsListPanel } = require('app/views/sidebar_sessions_list')
const { PlaylistListPanel } = require('app/views/sidebar_playlist_list')
const { PrecuePlayerView } = require("app/views/sidebar_precue_player")
const { AccordionLayout } = require("ui/layout/accordion")


class Sidebar extends AccordionLayout {
    constructor(library) {
        super()
        this.setStyle({
            width:"100%",
            height: "100%"
        })
        this.linksView = new LinksList(T)
        this.playlistsView = new PlaylistListPanel()
        this.playlistsView.displayModel(library.playlists)
        this.playlistsView.on("row-click", (playlistId) => {
            T.display_playlist(playlistId)
        })
        this.sessionsListView = new SessionsListPanel()
        this.sessionsListView.displayModel(library.sessions)
        this.sessionsListView.on("row-click", (sessionId) => {
            T.display_session(sessionId)
        })        
        this.precuePlayerView = new PrecuePlayerView(library.tracks)
        this.precuePlayerView.set_controller(AppController)

        this.addPanel("Library", [], this.linksView)
        this.addPanel("Playlists", [], this.playlistsView)
        this.addPanel("Sessions", [], this.sessionsListView)
        this.addPanel("Prelisten", [], this.precuePlayerView, 320)
    }
}

module.exports = {
    Sidebar: Sidebar
}
