// app/views/simple_track_list.js
const { SortableListView } = require('ui/listview/sortable')
const { CellRenderer } = require('ui/listview/renderer')
const $ = require("jquery")



class ElementRenderer extends CellRenderer {
    render(rowId, colId, obj) {
        let style = this._renderStyle(obj)
        let id = this._getValue("id", obj)
        if (id == null) {
            return `<td style="width:100%; margin:0; padding:0">
                <div class="queued-track element" style="background-color:rgb(153,50,204); line-height:50px; vertical-align:middle; text-align:center">
                    <b>END OF QUEUE</b>
                </div>
            </td>`
        } else {
            return `<td style="width:100%; margin:0; padding:0">
                <div class="queued-track element">
                    <img class="cover-image" src="${obj.cover}" height='50' width='50' draggable=false></img>
                    <div style="margin:0px; padding:0px; float:left; width:225px;">
                        <div class="queue_element_title"><b>${obj.title}</b></div>
                        <div class="queue_element_artist"><i>${obj.artist}</i></div>
                    </div>
                    <div style="float:right; width:65px; margin-right:10px; transform: translateY(30%);">
                        <div class="queue_element_bpm">${obj.bpm} BPM</div>
                        <div class="queue_element_duration">
                            <b>${obj.duration}</b>
                        </div>
                    </div>
                </div>
            </td>`
        }
    }
}


class SimpleTrackListView extends SortableListView {
    constructor(scrollerElement, tableBody) {
        super(scrollerElement, tableBody)
        this.tableBody = tableBody
        this.addColumnRenderer(new ElementRenderer({
            id: (obj) => {return obj.id},

            })
        )
    }

    convert_track(x) {
        let element = (x == null) ? {_id: null, duration: 0} : {
            id: x._id,
            title: x.metadata.title,
            artist: x.metadata.artist,
            bpm: x.track.bpm,
            duration: x.track.stream_end - x.track.stream_start,
        }
        if (element.id != null) {
            if (x.metadata.cover == null) {
                element.cover = "../../resources/images/default_album_cover.png"
            } else {
                element.cover = `file://${x.metadata.cover.small}`;
            }
        }
        return element
    }

    listReordered(new_order) {
        console.log(new_order)
    }

    setList(list_elements) {
        if (list_elements == undefined) {
            list_elements = []
        }
        let rowIds = this.setRows(list_elements.map((e) => {
            return this.convert_track(e)
        }))
        this.tableBody.innerHTML = this._rendered_rows.getList().join("")
        this._objectid_to_row_id = {}

        Object.keys(rowIds).forEach((id) => {
            this._objectid_to_row_id[rowIds[id].id] = id
        })
    }

    async move_selection_up() {
        let selected = await this.selected_element()
        let selected_id;
        if (selected !== undefined) {
            selected_id = (selected != null) ? selected._id : selected
            let position = this.view_list_order.indexOf(selected_id)
            if (position > 0) {
                let x = this.view_list_order[position - 1]
                this.view_list_order[position - 1] = selected_id
                this.view_list_order[position] = x
                await this.controller.reorder_queue(this.view_list_order)
                this.sortable.sort(this.view_list_order)
                this.num_tracks_dom.innerHTML = `${await this.controller.q_length()} tracks`
                this.duration_dom.innerHTML = `${format_seconds_long(Math.round(await this.controller.duration() / 1000))}`
            }
        }
    }

    async move_selection_down() {
        let selected = await this.selected_element()
        let selected_id;
        if (selected !== undefined) {
            selected_id = (selected != null) ? selected._id : selected
            let position = this.view_list_order.indexOf(selected_id)

            if (position + 1 < this.view_list_order.length) {
                let x = this.view_list_order[position + 1]
                this.view_list_order[position + 1] = selected_id
                this.view_list_order[position] = x
                await this.controller.reorder_queue(this.view_list_order)
                this.sortable.sort(this.view_list_order)
                this.num_tracks_dom.innerHTML = `${await this.controller.q_length()} tracks`
                this.duration_dom.innerHTML = `${format_seconds_long(Math.round(await this.controller.duration() / 1000))}`
            }
        }
    }

    handleDoubleClick(e) {
        let tableRow = e.target.closest("tr")
        let trackObject = this._elements.getElementById(tableRow.id)
        this.dispatch("row-double-click", trackObject.id)
    }

    handleDragEnter(e) {

    }

    handleDragEnd(e) {

    }

    handleDragOver(evt) {
        if (evt.preventDefault) {
            evt.preventDefault();
        }
        evt.dataTransfer.dropEffect = 'move';
    }

    handleDrop(evt) {
        if (evt.stopPropagation) {
            evt.stopPropagation();
        }
        let track_id = evt.dataTransfer.getData("text/plain")
        // let track = await MDB.getTrackById(track_id)
        // if (track) {
        //     this.controller.append(track)
        // }
    }
}

module.exports = {
    SimpleTrackListView: SimpleTrackListView
}
