// ui/dom/labeled_input.js
const { TextInput } = require("ui/dom/text_input.js")
const { Label } = require("ui/dom/label")
const { BoxLayout } = require("ui/layout/box")


class LabeledInput extends BoxLayout {
    constructor(label) {
        super("vertical", {left:0, right:0, top:0, bottom:0, padding:0})
        this.label = new Label(label)
        this.label.setStyle({
            fontSize: "9pt",
            textAlign: "left",
            textTransform: "uppercase",
            color: "rgb(30,30,30)",
            borderTop: "1px solid grey"
        })
        this.input = new TextInput()
        // this.input.domElement.classList.remove('main-track-search')
        // this.input.setStyle({
        //     marginTop: '8px',
        //     paddingLeft:'15px',
        //     color: 'rgb(20,20,20)',
        //     backgroundColor: "rgb(190,190,190)",
        //     fontSize: '10pt',
        //     border: 'none',
        //     outline: "none"
        // })
        this.append(this.input, true, true)
        this.append(this.label, true, true)
    }

    setText(v) {
        this.input.setText(v)
    }

    getText() {
        return this.input.getText()
    }
}


module.exports = {
    LabeledInput: LabeledInput
}
