// app/dialogs/automix_settings_dialog.js
const { InputDialog } = require("ui/dialog/input")
const { DOMElement } = require("ui/dom/domelement.js")
const { TextInput } = require("ui/dom/text_input.js")
const { Label } = require("ui/dom/label")
const { BoxLayout } = require("ui/layout/box")
const { LabeledInput } = require("ui/dom/labeled_input")

class AutomixSettingsDialog extends InputDialog {
    constructor(config) {
        super({
            confirmAction: config.confirmAction,
            dismissAction: config.dismissAction,
            confirmText: "save",
            dismissText: "cancel"
        })
        this.setTitle("Automix settings")
        this.setStyle({
            height: "200px",
            width:"300px"
        })
        this.name_input = this.makeInput("Seconds to wait between tracks...")
        this.input_area.append(this.name_input, true, true)

    }

    makeInput(placeholder) {
        let new_input = new LabeledInput("Gap between tracks (in milliseconds)")
        new_input.domElement.classList.remove('main-track-search')

        new_input.setStyle({
            width:'calc(100% - 20px)', 
            margin:'2px',
            // paddingLeft:'10px', 
            // paddingRight:'10px',
            // height:'20px',
            // color: "rgb(30,30,30)"
        })
        new_input.domElement.type = "text"
        return new_input
    }

}

module.exports = {
    AutomixSettingsDialog: AutomixSettingsDialog
}
