// app/dialogs/session_save_dialog.js
const { InputDialog } = require("ui/dialog/input")
const { DOMElement } = require("ui/dom/domelement.js")

class SessionSaveDialog extends InputDialog {
    constructor(config) {
        super({
            confirmAction: config.confirmAction,
            dismissAction: config.dismissAction,
            confirmText: "save",
            dismissText: "cancel"
        })
        this.setTitle("Save the current session...")
        this.setStyle({
            height: "150px",
            width:"300px"
        })
        this.name_input = this.makeInput("Name...")
        this.input_area.appendChildren([this.name_input])

    }

    makeInput(placeholder) {
        let new_input = new DOMElement("input")
        new_input.setStyle({
            width:'calc(100% - 20px)', 
            margin:'2px',
            paddingLeft:'10px', 
            paddingRight:'10px',
            height:'20px',
            color: "rgb(30,30,30)"
        })
        new_input.domElement.type = "text"
        new_input.domElement.placeholder = placeholder
        return new_input
    }

}

module.exports = {
    SessionSaveDialog: SessionSaveDialog
}
