// app/dialogs/session_save_dialog.js
const { InputDialog } = require("ui/dialog/input")
const { DOMElement } = require("ui/dom/domelement.js")
const { TextInput } = require("ui/dom/text_input.js")
const { Label } = require("ui/dom/label")
const { BoxLayout } = require("ui/layout/box")
const { LabeledInput } = require("ui/dom/labeled_input")


class SessionSaveDialog extends InputDialog {
    constructor(config) {
        super({
            confirmAction: config.confirmAction,
            dismissAction: config.dismissAction,
            confirmText: "save",
            dismissText: "cancel"
        })
        this.setTitle("Save the current session...")
        this.setStyle({
            height: "200px",
            width:"300px"
        })
        this.name_input = this.makeInput("Name...")
        this.input_area.appendChildren([this.name_input])

    }

    makeInput(placeholder) {
        let new_input = new LabeledInput("Session name")
        new_input.setStyle({
            width:'calc(100% - 20px)', 
            margin:'2px'
        })
        new_input.domElement.type = "text"
        return new_input
    }

}

module.exports = {
    SessionSaveDialog: SessionSaveDialog
}
