// app/dialogs/session_save_dialog.js
const { InputDialog } = require("ui/dialog/input")
const { DOMElement } = require("ui/dom/domelement.js")
const { TextInput } = require("ui/dom/text_input.js")
const { Label } = require("ui/dom/label")
const { BoxLayout } = require("ui/layout/box")


class LabeledInput extends BoxLayout {
    constructor(label) {
        super("vertical", {left:0, right:0, top:0, bottom:0, padding:0})
        this.label = new Label(label)
        this.label.setStyle({
            fontSize: "9pt",
            textAlign: "left",
            textTransform: "uppercase",
            color: "rgb(30,30,30)",
            borderTop: "1px solid grey"
        })
        this.input = new TextInput()
        this.input.domElement.classList.remove('main-track-search')
        this.input.setStyle({
            marginTop: '8px',
            paddingLeft:'15px',
            color: 'rgb(20,20,20)',
            backgroundColor: "rgb(190,190,190)",
            fontSize: '10pt',
            border: 'none',
            outline: "none"
        })
        this.append(this.input, true, true)
        this.append(this.label, true, true)
    }

    setText(v) {
        this.input.setText(v)
    }

    getText() {
        return this.input.getText()
    }
}

class SessionSaveDialog extends InputDialog {
    constructor(config) {
        super({
            confirmAction: config.confirmAction,
            dismissAction: config.dismissAction,
            confirmText: "save",
            dismissText: "cancel"
        })
        this.setTitle("Save the current session...")
        this.setStyle({
            height: "150px",
            width:"300px"
        })
        this.name_input = this.makeInput("Name...")
        this.input_area.appendChildren([this.name_input])

    }

    makeInput(placeholder) {
        let new_input = new LabeledInput("Session name")
        new_input.setStyle({
            width:'calc(100% - 20px)', 
            margin:'2px',
            paddingLeft:'10px', 
            paddingRight:'10px',
            height:'20px',
            color: "rgb(30,30,30)"
        })
        new_input.domElement.type = "text"
        return new_input
    }

}

module.exports = {
    SessionSaveDialog: SessionSaveDialog
}
