// app/dialogs/track_metadata_editor.js

const { RemoteTrackPlayer } = require("track_playback/remote_player/remote_player.js")
const { DOMElement } = require("ui/dom/domelement.js")
const { Image } = require("ui/dom/image.js")
const { TextInput } = require("ui/dom/text_input.js")
const { Label } = require("ui/dom/label")
const { Dialog } = require("ui/dialog/dialog.js")
const { BoxLayout } = require("ui/layout/box")
const { ConfirmButton, DismissButton } = require("ui/dialog/buttons.js")
const moment = require('moment');


class Rating extends DOMElement {
    constructor() {
        super('div')
        this.setRating(0)
    }

    clear() {
        while (this.domElement.firstChild) {
            this.domElement.removeChild(this.domElement.firstChild);
        }
        this.stars = []
    }

    setRating (num) {
        this.rating = num
        this.clear()
        for (let i=1; i<6; i++) {
            let star = document.createElement("i")
            star.classList.add('fa')
            star.classList.add(`${( i <= num ? "fa-star" : "fa-star-o")}`)
            star.style.width = '11px'
            star.style.color = "rgb(30,30,30)"
            star.style.marginLeft = i > 1 ? "3px" : "0px"
            this.domElement.appendChild(star)
            this.stars.push(star)
        }
        for (let j=0; j<this.stars.length; j++) {
            this.stars[j].addEventListener('click', () => {
                if (j == 0 && this.rating == 1) {
                    this.setRating(0)
                } else {
                    this.setRating(j+1)
                }
            })

        }
    }

    getRating() {
        return this.rating
    }
}

class CustomCheckbox extends DOMElement {
    constructor() {
        super('div')
        this.setValue(false)
    }

    clear() {
        while (this.domElement.firstChild) {
            this.domElement.removeChild(this.domElement.firstChild);
        }
    }

    setValue (v) {
        this.checked = !!v
        this.clear()
        let check = document.createElement("i")
        check.classList.add('fa')
        check.classList.add(`${( this.checked ? "fa-heart" : "fa-heart-o")}`)
        check.style.width = '11px'
        check.style.color = "rgb(30,30,30)"
        this.domElement.appendChild(check)
        check.addEventListener('click', () => {
            this.setValue(!this.checked)
        })

    }

    getValue() {
        return this.checked
    }
}


class LabelledInput extends BoxLayout {
    constructor(label) {
        super("vertical", {left:0, right:0, top:0, bottom:0, padding:0})
        this.label = new Label(label)
        this.label.setStyle({
            fontSize: "9pt",
            textAlign: "left",
            textTransform: "uppercase",
            color: "rgb(30,30,30)",
            borderTop: "1px solid grey"
        })
        this.input = new TextInput()
        this.input.domElement.classList.remove('main-track-search')
        this.input.setStyle({
            marginTop: '8px',
            paddingLeft:'15px',
            color: 'rgb(20,20,20)',
            backgroundColor: "rgb(190,190,190)",
            fontSize: '10pt',
            border: 'none',
            outline: "none"
        })
        this.append(this.input, true, true)
        this.append(this.label, true, true)
    }

    setText(v) {
        this.input.setText(v)
    }

    getText() {
        return this.input.getText()
    }
}

class TrackMetadataDialog extends Dialog {
    constructor(track, library, config) {
        super()
        this.setTitle("Track info")
        this.setStyle({width:'500px', height:"850px"})
        this.library = library
        this.track = track

        this.audio_player = new RemoteTrackPlayer()
        this.audio_player.addOutput("headphones")
        this.setOutputDeviceId("null")
        this.audio_player.initialize()

        this.coverImage = new Image()
        this.coverImage.setStyle({
            border: "1px solid black"
        })
        this.titleLabel = new Label("title")
        this.titleLabel.setStyle({
            color: 'rgb(30,30,30)',
            fontWeight: "bold",
            fontSize: "12pt"
        })
        this.artistLabel = new Label("artist")
        this.artistLabel.setStyle({
            color: 'rgb(30,30,30)',
            fontSize: "10pt"
        })

        this.albumLabel = new Label("album")
        this.albumLabel.setStyle({
            color: 'rgb(30,30,30)',
            fontSize: "10pt"
        })

        this.ratingInput = new Rating()
        this.favoriteInput = new CustomCheckbox()
        let ratingLayout = new BoxLayout("horizontal", {left:0, right:0, top:0, bottom:0, padding:10})
        ratingLayout.append(this.ratingInput, false, true, 75)
        ratingLayout.append(this.favoriteInput, false, true, 15)

        let layout = new BoxLayout("horizontal", {left:0, right:0, top:0, bottom:0, padding:10})
        let metadataLayout = new BoxLayout("vertical", {left:0, right:0, top:0, bottom:0, padding:5})
        metadataLayout.append(this.titleLabel, false, true, 20)
        metadataLayout.append(this.artistLabel, false, true, 20)
        metadataLayout.append(this.albumLabel, false, true, 20)
        metadataLayout.append(ratingLayout, false, true, 30)
        layout.append(this.coverImage, false, true, 125)
        layout.append(metadataLayout, true, true)


        this.titleInput = new LabelledInput("title")
        this.artistInput = new LabelledInput("artist")
        this.albumInput = new LabelledInput("album")
        this.genreInput = new LabelledInput("genre")
        this.yearInput = new LabelledInput("year")
        this.bpmInput = new LabelledInput("bpm")
        let genreLayout = new BoxLayout("horizontal", {left:0, right:0, top:0, bottom:0, padding:10})
        genreLayout.append(this.genreInput, true, true)
        genreLayout.append(this.yearInput, true, true)
        genreLayout.append(this.bpmInput, true, true)
        // genreLayout.append(this.ratingInput, false, true, 75)
        // genreLayout.append(this.favoriteInput, false, true, 15)


        this.layout = new BoxLayout("vertical", {left:20, right:20, top:20, bottom:20, padding:10})
        this.layout.setStyle({
            position: "absolute",
            top: '0px',
            width: "100%",
            height: "100%"
        })


        let save_button = new ConfirmButton()
        save_button.setText("Save")
        save_button.on("click", async () => {
            this.audio_player.stop()
            console.log(this.getValues())
            this.close()
        })

        let dismiss_button = new DismissButton()
        dismiss_button.setText("Close")
        dismiss_button.on("click", () => {
            this.audio_player.stop()
            this.close()
        })

        let button_layout = new BoxLayout("horizontal", {padding: 20})
        button_layout.append(save_button, true, true)
        button_layout.append(dismiss_button, true, true)

        this.layout.append(layout, false, true, 125)
        this.layout.append(this.titleInput, false, true, 40)
        this.layout.append(this.artistInput, false, true, 40)
        this.layout.append(this.albumInput, false, true, 40)
        this.layout.append(genreLayout, false, true, 40)
        this.layout.append(new DOMElement("div"), true, true)
        this.layout.append(button_layout, false, true, 50)
        this.appendChildren([this.layout])
        this.setTrack(track)
    }

    setOutputDeviceId(deviceId) {
        (deviceId != this.getOutputDeviceId()) && this.audio_player.setOutputDeviceId("headphones", deviceId)
    }

    getOutputDeviceId() {
        return this.audio_player.getOutputDeviceIds()["headphones"]
    }

    async open() {
        super.open()
    }

    getValues() {
        let v = {
            "metadata.title": this.titleInput.getText(), 
            "metadata.album": this.albumInput.getText(),
            'metadata.artist': this.artistInput.getText(), 
            "track.bpm": this.bpmInput.getText(), 
            "metadata.genre": this.genreInput.getText(),
            "metadata.year": this.yearInput.getText(), 
            "stats.loved": this.favoriteInput.getValue() || false,
            "stats.rating": this.ratingInput.getRating() || 0
        }
        v.cover_image = this.cover_image
        v['metadata.year'] = (v['metadata.year'] != "") ? v['metadata.year'] : null
        v['track.bpm'] = (v['track.bpm'] != "") ? v['track.bpm'] : null
        return v
    }


    async setTrack(track) {
        this.track = track
        this.titleLabel.setText(track.metadata.title)
        this.albumLabel.setText(track.metadata.album)
        this.artistLabel.setText(track.metadata.artist)
        this.titleInput.setText(track.metadata.title)
        this.albumInput.setText(track.metadata.album)
        this.artistInput.setText(track.metadata.artist)
        this.bpmInput.setText(track.track.bpm)
        this.genreInput.setText(track.metadata.genre)
        this.yearInput.setText(track.metadata.year)
        this.ratingInput.setRating(track.stats.rating)
        this.favoriteInput.setValue(track.stats.favorite)

        let cover_source = undefined
        if (track.metadata.cover == null) {
            cover_source = "../../resources/images/default_album_cover.png"
            this.original_cover_image = null
            this.cover_image = null
        } else {
            cover_source = `file://${track.metadata.cover.original}`;
            this.original_cover_image = cover_source
            this.cover_image = undefined
        }
        this.coverImage.setImage(cover_source)
    }
}

module.exports = {
    TrackMetadataDialog: TrackMetadataDialog
}
