// track_playback/track_player.js
const {PydjayAudioFilePlayer} = require("webaudio/audio_player_file.js")
const $ = require('jQuery');
class AudioTrackPlayer extends PydjayAudioFilePlayer {
    constructor() {
        super()
        this._current_track = undefined
        this.on('end-of-stream', () => {
            if (this._current_track != undefined) {
                this._current_track.time_end = new Date()
                this._current_track.status = "ended"
                this.dispatch("track-finished", this._current_track)
                this._current_track = undefined    
            }
        })
    }

    play(track, stream_start, stream_end) {
        let file_name = track.track.path

        if (stream_start == undefined) {
            stream_start = track.track.stream_start
            stream_end = track.track.stream_end
        } else if (stream_end == undefined) {
            stream_end = track.track.stream_end
            if (stream_start < 0) {
                stream_start = stream_end + stream_start;
            }
        }

        super.play(file_name, stream_start, stream_end)
        this._current_track = {
            track_object: track,
            time_start: new Date(),
            time_end: undefined,
            status: undefined
        }
        this.dispatch("track-started", this._current_track)
    }

    playUrl(url, stream_start, stream_end) {
        super.play(url, stream_start, stream_end)
    }

    stop() {
        if (this._current_track != undefined) {
            this._current_track.time_end = new Date()
            this._current_track.status = "stopped"
            this.dispatch("track-stopped", this._current_track)
            this._current_track = undefined
        }
        super.stop()
    }
}

module.exports = {
    AudioTrackPlayer: AudioTrackPlayer
}
