// track_playback/track_player.js
const {PydjayAudioFilePlayer} = require("webaudio/audio_player_file.js")

class AudioTrackPlayer extends PydjayAudioFilePlayer {
    constructor() {
        super()
        this._current_track = undefined
        this.on('end-of-stream', () => {
            this._current_track.end_time = new Date()
            this._current_track.status = "ended"
            this.dispatch("track-finished")
            this._current_track = undefined
        })
    }

    reset_audio_context() {
        let url = undefined,
            time = undefined,
            end_time = undefined;

        if (this.source != undefined) {
            time = this.source.currentTime * 1000
            end_time = this.stream_end
            url = this.url
            this.stop()
        }

        super.reset_audio_context()
        this.init_audio()

        if (url != undefined) {
            super.play(url, time, end_time)
        }

        this.dispatch("audio-system-reset")
    }

    play(track, stream_start, stream_end) {
        this.current_track = {
            track_object: track,
            time_start: new Date(),
            time_end: undefined,
            status: undefined
        }

        let file_name = path.join(track.music_root, track.file_name);

        if (stream_start == undefined) {
            stream_start = track.stream_start
            stream_end = track.stream_end
        } else if (stream_end == undefined) {
            stream_end = track.stream_end
            if (stream_start < 0) {
                stream_start = stream_end + stream_start;
            }
        }
        this.dispatch("track-started", this._current_track)
        super.play(file_name, stream_start / 1000000, stream_end / 1000000)
    }

    stop() {
        this._current_track.end_time = new Date()
        this._current_track.status = "stopped"
        this.dispatch("track-stopped", this._current_track)
        this._current_track = undefined
    }
}

module.exports = {
    AudioTrackPlayer: AudioTrackPlayer
}
