// ui/dialog/dialog.js

// import DOMElement from "../dom/domelement.js"
const {DOMElement} = require("ui/dom/domelement.js")

class Dialog extends DOMElement {
    constructor () {
        super("dialog")
        this._titleElement = new DOMElement("h1")
        this._titleElement.setStyle({
            backgroundColor: "rgb(30,30,30)", 
            textAlign:  "center", 
            fontSize:   "11pt", 
            margin:     "0px",
            color:      "rgb(200,200,200)", 
            height:     "30px", 
            lineHeight: "30px"
        })
        this.domElement.appendChild(this._titleElement.domElement)
        this.setStyle({
            zIndex:  1000, 
            padding: 0, 
            backgroundColor: "rgb(200,200,200)",
            border: "2px solid grey"
        })
    }

    setTitle(title) {
        this._titleElement.domElement.innerHTML = title
    }

    open() {
        if (!this.domElement.open) {
            let window = document.getElementsByTagName("body")[0]
            window.appendChild(this.domElement)
            this.domElement.showModal()    
        }
    }

    close() {
        if (this.domElement.open) {
            let window = document.getElementsByTagName("body")[0]
            this.domElement.close()
            window.removeChild(this.domElement)    
        }
    }
}

module.exports = {
    Dialog: Dialog
}
