// ui/dialog/dialog.js

const {DOMElement} = require("ui/dom/domelement.js")

class Dialog extends DOMElement {
    constructor () {
        super("dialog")
        this._titleElement = new DOMElement("h1")
        // this._titleElement.setStyle({
        //     backgroundColor: "rgb(30,30,30)", 
        //     textAlign: "center", 
        //     fontSize: "11pt", 
        //     margin: "0px",
        //     color: "rgb(200,200,200)", 
        //     height: "30px", 
        //     lineHeight: "30px"
        // })
        this.domElement.appendChild(this._titleElement.domElement)
        this.setStyle({
            zIndex: 1000, 
            padding: 0, 
            backgroundColor: "rgb(200,200,200)",
            border: "2px solid grey"
        })
        this._bodyElement = new DOMElement("div")
        this._bodyElement.setStyle({
            position: "absolute",
            top: "30px",
            height:"calc(100% - 30px)",
            width: "100%"
        })
        this.domElement.appendChild(this._bodyElement.domElement)
    }

    appendChildren(x) {
        x.forEach((e) => {
            this._bodyElement.domElement.appendChild(e.domElement)
        })
    }

    setTitle(title) {
        this._titleElement.domElement.innerHTML = title
    }

    open() {
        if (!this.domElement.open) {
            let window = document.getElementsByTagName("body")[0]
            window.appendChild(this.domElement)
            this.domElement.showModal()    
        }
    }

    close() {
        if (this.domElement.open) {
            let window = document.getElementsByTagName("body")[0]
            this.domElement.close()
            window.removeChild(this.domElement)    
        }
    }

}

module.exports = {
    Dialog: Dialog
}
