// ui/dialog/question.js

const {DOMElement} = require("ui/dom/domelement.js")
const {Dialog} = require("./dialog.js")
const { InputDialog } = require("./input")
const {ConfirmButton, DismissButton} = require("./buttons.js")
const { BoxLayout } = require("ui/layout/box")

const path = require("path")

class Question extends InputDialog {
    constructor (config) {
        super(config)
        // this._config = config
        this.setStyle({width: "450px", height: "200px"})
        // this.setTitle(config.title)
        this._icon = new DOMElement("img")
        this._icon.setStyle({
            width:  "64px",
            height: "64px",
            margin: "20px",
            display: "inline-block",
            float: "left"})
            this._icon.domElement.src = path.join(__dirname, "resources/icon-warning.png")

        let d = new DOMElement("div")
        d.setStyle({
            color:   "rgb(30,30,30)",
            width:   "calc(100% - 104px)",
            height:  "84px",
            display: "inline-block",
            padding: 0,
            marginTop: "20px"})

        this._question_text = new DOMElement("div")
        this._question_text.setStyle({
            color: "rgb(30,30,30)",
            width: "calc(100% - 20px)",
            display: "inline-block",
            textAlign: "justify"})
        this._question_text.domElement.innerHTML = config.question
        d.appendChild(this._question_text)

        let q = new BoxLayout("horizontal", {padding:30})
        q.append(this._icon, false, true, 64)
        q.append(d, true, true)
        this.input_area.append(q, true, true)
    }
}

module.exports = {
    Question: Question
}
