// ui/dialog/question.js

const {DOMElement} = require("ui/dom/domelement.js")
const {Dialog} = require("./dialog.js")
const {ConfirmButton, DismissButton} = require("./buttons.js")
const path = require("path")

class Question extends Dialog {
    constructor (config) {
        super()
        this._config = config
        this.setStyle({width: "450px", height: "200px"})
        this.setTitle(config.title)
        this._icon = new DOMElement("img")
        this._icon.setStyle({
            width:  "64px",
            height: "64px",
            margin: "20px",
            display: "inline-block",
            float: "left"})
            this._icon.domElement.src = path.join(__dirname, "resources/icon-warning.png")

        let d = new DOMElement("div")
        d.setStyle({
            color:   "rgb(30,30,30)",
            width:   "calc(100% - 104px)",
            height:  "84px",
            display: "inline-block",
            padding: 0,
            marginTop: "20px"})

        this._question_text = new DOMElement("div")
        this._question_text.setStyle({
            color: "rgb(30,30,30)",
            width: "calc(100% - 20px)",
            display: "inline-block",
            textAlign: "justify"})
        this._question_text.domElement.innerHTML = config.question
        d.appendChild(this._question_text)

        let q = new DOMElement("div")
        q.appendChildren([this._icon, d])

        this._confirm_button = new ConfirmButton()
        this._confirm_button.setStyle({width:"calc(50% - 16px)"})
        this._confirm_button.setText(config.confirmText || "confirm")

        this._dismiss_button = new DismissButton()
        this._dismiss_button.setText(config.dismissText || "dismiss")
        this._dismiss_button.setStyle({width:"calc(50% - 16px)"})

        d = new DOMElement("div")
        d.setStyle({width:"100%", height:"50px"})

        let button_area = new DOMElement("div")
        button_area.setStyle({
            width:"55%",
            height: "50px",
            left:"25%",
            position:"relative"})
        button_area.appendChildren([this._confirm_button, this._dismiss_button])

        d.appendChild(button_area)

        this.appendChildren([q, d])

        this._confirm_button.on("click", () => {
            if (this._config.confirmAction != undefined) {
                this._config.confirmAction(this)
            }
        })

        this._dismiss_button.on("click", () => {
            if (this._config.dismissAction != undefined) {
                this._config.dismissAction(this)
            }
        })
    }
}

module.exports = {
    Question: Question
}
