// app/components/list_header.js
const { Button } = require("ui/dom/button.js")
const { Label } = require("ui/dom/label")
const { TextInput } = require("ui/dom/text_input.js")
const { TagFilterPicker } = require("ui/popup/tag_filter_picker")
const { BoxLayout } = require("ui/layout/box")


class ListInfoHeader extends BoxLayout {
    constructor(library, filtering) {
        super("horizontal", {left: 5, right:5, top:5, bottom:5, padding:10})
        this.library = library
        this.filtering = filtering
        this.filterUnplayed = false
        this.filterTags = []

        this.listNameLabel = new Label()
        this.listStatLabel = new Label()

        let nameLayout = new BoxLayout("vertical", {padding:5})
        nameLayout.append(this.listNameLabel, true, true)
        nameLayout.append(this.listStatLabel, true, true)

        this.append(nameLayout, true, true)

        if (this.filtering) {
            this.tagFilterPicker = undefined
            this.tagFilterButton = new Button()
            this.filterInput = new TextInput()
            let filterLayout = new BoxLayout("horizontal", {padding:0})
            filterLayout.append(this.tagFilterButton, false, true, 25)
            filterLayout.append(this.filterInput, true, true)
            this.append(filterLayout, false, true, 100)
            this.initTextFilter()
            this.initTagFilter()
        }
    }

    initTextFilter() {
        this.filterInput.domElement.addEventListener("keyup", (e) => {
            if ((e.key == "Escape") || (e.key == "Enter")) {
                this.filterInput.domElement.blur()
            }
        });
        this.filterInput.domElement.oninput = (e) => {
            this.dispatch("filter-changed", this.filterInput.domElement)
        }
    }

    initTagFilter() {
        this.tagFilterButton.on("click", async () => {
            let tags = Object.values(await this.library.tags.getAllObjects()).map((x) => {
                return {
                    id: x._id,
                    name: x.name,
                    checked: (this.filterTags[x._id] != undefined) 
                }
            })
            if (!this.tagFilterPicker) {
                this.tagFilterPicker = new TagFilterPicker(document.getElementsByTagName("body")[0], {
                    filterUnplayed: this.filterUnplayed,
                    tags: tags, 
                    cancel: () => {
                        this.tagFilterPicker = undefined
                    },
                    updateFilter: async (tags) => {
                        this.tagFilterPicker = undefined
                        this.filterUnplayed = tags.filterUnplayed
                        this.filterTags = tags.filterTags
                        this.dispatch("tag-filter-changed", tags)
                    }                    
                })
                this.tagFilterPicker.open(this.tagFilterButton.domElement)    
            } else {
                this.tagFilterPicker.close()
                this.tagFilterPicker = undefined
            }
        })

    }

    setMetadata(metadata) {
        this.listNameLabel.setText(metadata.name)
        this.listStatLabel.setText(`${metadata.numTracks} tracks - ${format_seconds_long(metadata.duration / 1000)}`)
    }
}

module.exports = {
    ListInfoHeader: ListInfoHeader
}
