// app/components/rating.js
const { DOMElement } = require("ui/dom/domelement.js")


class Rating extends DOMElement {
    constructor(color) {
        super('div')
        this.setRating(0)
        this.color = color || "rgb(30,30,30)"
    }

    clear() {
        while (this.domElement.firstChild) {
            this.domElement.removeChild(this.domElement.firstChild);
        }
        this.stars = []
    }

    setRating (num) {
        this.rating = num
        this.clear()
        for (let i=1; i<6; i++) {
            let star = document.createElement("i")
            star.classList.add('fa')
            star.classList.add(`${( i <= num ? "fa-star" : "fa-star-o")}`)
            star.style.width = '11px'
            star.style.color = this.color
            star.style.marginLeft = i > 1 ? "3px" : "0px"
            this.domElement.appendChild(star)
            this.stars.push(star)
        }
        
        for (let j=0; j<this.stars.length; j++) {
            this.stars[j].addEventListener('click', () => {
                if (j == 0 && this.rating == 1) {
                    this.setRating(0)
                    this.dispatch("changed", 0)
                } else {
                    this.setRating(j+1)
                    this.dispatch("changed", j+1)
                }
            })
        }
    }

    getRating() {
        return this.rating
    }
}

module.exports = {
    Rating: Rating
}
