// ui/layout/box.js
const { DOMElement } = require("ui/dom/domelement.js")


class BoxLayout extends DOMElement {
    constructor(orientation, margins) {
        super("div")
        this.orientation = orientation || "vertical"
        this._blockMode = orientation == 'vertical' ? "block" : 'inline-block' 
        this.children = []
        this.domElement.addEventListener("resize", (event) => {
            if (this.orientation == 'horizontal') {
                this.layoutHeight = this._layout.domElement.clientWidth - paddingHeight
            } else if (this.orientation == 'vertical') {
                this.layoutHeight = this._layout.domElement.clientheight - paddingHeight
            } 
            this.layout()  
        })
        margins = margins || {left:0, right:0, top:0, bottom:0}
        this.paddingHeight = margins.padding || 0
        this._layout = new DOMElement("div")
        let marginSize = this._computeMargin(margins)
        this._layout.setStyle({
            position: "absolute",
            top: "0px",
            left: "0px",
            marginLeft: `${margins.left || 0}px`,
            marginRight: `${margins.right|| 0}px`,
            marginTop : `${margins.top || 0}px`,
            marginBottom : `${margins.bottom || 0}px`,
            width: `calc(100% - ${marginSize.horizontal}px)`,
            height: `calc(100% - ${marginSize.vertical}px)`
        })
        this.appendChild(this._layout)
        this.layoutObserver = new MutationObserver((...x) => {
            let paddingHeight = this.paddingHeight * (this.children.length - 1)
            if (this.orientation == 'horizontal') {
                this.layoutHeight = this._layout.domElement.clientWidth - paddingHeight
            } else if (this.orientation == 'vertical') {
                this.layoutHeight = this._layout.domElement.clientHeight - paddingHeight
            } 
            this.layout()
        })
        this.layoutObserver.observe(this.domElement, {
            childList: true,
            attributes: true,
            subtree: true
        })
    }

    _computeMargin(x) {
        let horizontal = x ? (x.left || 0) + (x.right || 0) : 0
        let vertical = x ? (x.top || 0) + (x.bottom || 0) : 0
        return {horizontal: horizontal, vertical: vertical}
    }

    _expandDimension(x) {
        if (this.orientation == "vertical") {
            x.setStyle({width: "100%"})
        } else if (this.orientation == "horizontal") {
            x.setStyle({height: "100%"})
        }
    }

    padding() {
        let x = new DOMElement('div')
        x.setStyle({
            display:this._blockMode
        })
        this._expandDimension(x)
        if (this.orientation == "vertical") {
            x.setStyle({height: this.paddingHeight+"px"})
        } else if (this.orientation == "horizontal") {
            x.setStyle({width: this.paddingHeight+"px"})
        }
        return x
    }

    append(child, expand, fill, dimension) {
        let container = new DOMElement("div")
        container.setStyle({
            position: "relative",
            display: this._blockMode,
        })

        if (this.orientation == 'vertical') {
            dimension && container.setStyle({height: `${dimension}px`})
        } else if (this.orientation == 'horizontal') {
            dimension && container.setStyle({width: `${dimension}px`})
        }

        this._expandDimension(container)

        child.setStyle({
            position: (dimension || fill) ? "absolute" : "relative",
            top: "0px",
            left: "0px",
            margin: "0px",
            padding: "0px",
            display: "block"
        })
        if (fill) {
            child.setStyle({
                width: "100%",
                height: "100%"
            })
        }

        container.appendChild(child);
        
        (this.children.length > 0) && this._layout.appendChild(this.padding())
        
        this._layout.appendChild(container)
        this.children.push({
            container: container,
            element: child,
            expand: (expand && true)  || false,
            fill: (fill && true)  || false
        })

    }

    layout() {
        let fixedHeight = 0
        let expandCount = 0
        this.children.forEach((x) => {
            if (!(x.expand)) {
                fixedHeight += (this.orientation == 'vertical') ? 
                                    x.container.domElement.clientHeight :
                                    x.container.domElement.clientWidth
                                    
            } else {
                expandCount += 1
            }
        })
        let expandHeight = Math.round((this.layoutHeight - fixedHeight) / expandCount)
        this.children.forEach((x) => {
            if (x.expand) {
                if (this.orientation == 'vertical') {
                    x.container.domElement.style.height = expandHeight+"px"
                } else if (this.orientation == 'horizontal') {
                    x.container.domElement.style.width = expandHeight+"px"
                }
            } 
        })
    }
}

module.exports = {
    BoxLayout: BoxLayout
}
