// ui/layout/vertical.js
const { DOMElement } = require("ui/dom/domelement.js")


class VerticalLayout extends DOMElement {
    constructor(margins) {
        super("div")
        this.children = []
        this.domElement.addEventListener("resize", (event) => {
            this.layoutHeight = this._layout.domElement.clientHeight
            this.layout()  
        })
        this.paddingHeight = margins.padding || 0
        this._layout = new DOMElement("div")
        let marginSize = this._computeMargin(margins)
        margins = margins || {left:0, right:0, top:0, bottom:0}
        this._layout.setStyle({
            position: "absolute",
            top: "0px",
            left: "0px",
            marginLeft: `${margins.left || 0}px`,
            marginRight: `${margins.right|| 0}px`,
            marginTop : `${margins.top || 0}px`,
            marginBottom : `${margins.bottom || 0}px`,
            width: `calc(100% - ${marginSize.horizontal}px)`,
            height: `calc(100% - ${marginSize.vertical}px)`
        })
        this.appendChild(this._layout)
        this.layoutObserver = new MutationObserver((...x) => {
            let paddingHeight = this.paddingHeight * (this.children.length - 1)
            this.layoutHeight = this._layout.domElement.clientHeight - paddingHeight
            this.layout()
        })
        this.layoutObserver.observe(this.domElement, {
            childList: true,
            attributes: true,
            subtree: true
        })
    }

    _computeMargin(x) {
        let horizontal = x ? (x.left || 0) + (x.right || 0) : 0
        let vertical = x ? (x.top || 0) + (x.bottom || 0) : 0
        return {horizontal: horizontal, vertical: vertical}
    }

    padding() {
        let x = new DOMElement('div')
        x.setStyle({
            width:'100%',
            height: this.paddingHeight+"px",
            display:"block"
        })
        return x
    }

    append(child, expand, fill) {
        let container = new DOMElement("div")
        container.setStyle({
            width: `calc(100% - 0px)`,
            position: "relative",
            display: "block",
        })
        child.setStyle({
            position: "relative",
            top: "0px",
            left: "0px",
            margin: "0px",
            padding: "0px",
            display: "block"
        })
        if (fill) {
            child.setStyle({
                width: "100%",
                height: "100%"
            })
        }
        container.appendChild(child);
        
        (this.children.length > 0) && this._layout.appendChild(this.padding())
        
        this._layout.appendChild(container)
        this.children.push({
            container: container,
            element: child,
            expand: (expand && true)  || false,
            fill: (fill && true)  || false
        })

    }

    layout() {
        let fixedHeight = 0
        let expandCount = 0
        this.children.forEach((x) => {
            if (!(x.expand)) {
                fixedHeight += x.container.domElement.clientHeight
            } else {
                expandCount += 1
            }
        })
        let expandHeight = Math.round((this.layoutHeight - fixedHeight) / expandCount)
        this.children.forEach((x) => {
            if (x.expand) {
                x.container.domElement.style.height = expandHeight+"px"
            } 
        })
    }
}

module.exports = {
    VerticalLayout: VerticalLayout
}
