// webaudio/audio_context.js
const {EventDispatcher} = require("notify/event_dispatcher.js")

class PydjayAudioContext extends EventDispatcher{
    constructor() {
        super()
        this.audio_ctx =  new AudioContext()
        this.destination = this.audio_ctx.createMediaStreamDestination()
        this.time_callback = null
        this.time_monitor = this.audio_ctx.createScriptProcessor(256, 1, 1)
        this.time_monitor.onaudioprocess = () => {
            this.dispatch("timestamp", this.audio_ctx.currentTime)
        }
        this.time_monitor.connect(this.destination)
        this.source = null;
        this.outputs = {}
        this._mediaGroup = '_' + Math.random().toString(36).substr(2, 9);
    }

    addOutput(name) {
        if (this.outputs[name] == undefined) {
            this.outputs[name] = new Audio()
            this.outputs[name].mediaGroup = this._mediaGroup
            this.outputs[name].src = URL.createObjectURL(this.destination.stream);
        }
    }
}

module.exports = {
    PydjayAudioContext:PydjayAudioContext
}
